File: Cores/ExtremumScalaCore/src/main/scala/OSOL/Extremum/Core/Scala/Optimization/Testing/Tester.scala
package OSOL.Extremum.Core.Scala.Optimization.Testing

import OSOL.Extremum.Core.Scala.Optimization.{Algorithm, Area, Optimizable}
import OSOL.Extremum.Core.Scala.Vectors.{RealVector, VectorObject}

abstract class Tester[Base, FuncType, V <: Optimizable[Base, FuncType]](testFunctions: Seq[Map[String, FuncType] => FuncType],
                                                                        areas: Seq[Area],
                                                                        solutions: Seq[Map[String, Double]],
                                                                        tolerance: Double,
                                                                        attempts: Int) {

  def Lp_norm(v1: VectorObject[Double], v2: VectorObject[Double], p: Int = 2): Double = {
    val keys = v1.keys ++ v2.keys
    keys.map(k => math.pow(math.abs(v1(k) - v2(k)), p)).sum
  }

  def apply(algorithms: Seq[Algorithm[Base, FuncType, V]]): Boolean = {
    val resultsPerFunction = testFunctions.zip(areas).zip(solutions)
      .map { case ((f, area), sol) => {
        var success = false
        val resultsPerAlgorithm = algorithms.takeWhile { a => {
          val resultsPerAttempt = (1 to attempts).takeWhile { _ =>
            val r = a.work(f, area).toBasicForm()
            success = Lp_norm(r, RealVector(sol)) < tolerance
            !success
          }
          !success
        }
        }
        success
      }
      }
    resultsPerFunction.forall(r => r)
  }

}

File: Cores/DotNet/Optimization/Testing/Tester.cs
using System;
using System.Collections.Generic;
using System.Linq;
using OSOL.Extremum.Cores.DotNet.Optimization.RemoteFunctions;
using OSOL.Extremum.Cores.DotNet.Vectors;

namespace OSOL.Extremum.Cores.DotNet.Optimization.Testing
{
    public abstract class Tester<TBase, TFuncType, TV> where TV : class, IOptimizable<TBase, TFuncType>
    {
        public RemoteFunction<TFuncType>[] TestFunctions;
        public Dictionary<string, Tuple<double, double>>[] Areas;
        public Dictionary<string, double>[] Solutions;
        public double Tolerance;
        public int Attempts;

        public double Lp_norm(VectorObject<double> v1, VectorObject<double> v2, int p = 2)
        {
            var keys = v1.Keys.ToList();
            foreach (var key in v2.Keys)
            {
                if (!keys.Contains(key))
                {
                    keys.Add(key);
                }
            }
            return keys.Select(k => Math.Pow(Math.Abs(v1[k] - v2[k]), p)).Sum();
        }

        public bool Check(params Algorithm<TBase, TFuncType, TV>[] algorithms)
        {
            bool[] resultsPerFunction = new bool[TestFunctions.Length];
            for (int id = 0; id < TestFunctions.Length; ++id)
            {
                var f = TestFunctions[id];
                var area = Areas[id];
                var sol = Solutions[id];
                bool success = false;
                for (int alg_id = 0; alg_id < algorithms.Length && !success; ++alg_id)
                {
                    var algorithm = algorithms[alg_id];
                    for (int attempt = 0; attempt < Attempts && !success; ++attempt)
                    {
                        algorithm.Reset();
                        f.Initialize();
                        var r = algorithm.Work(f.Calculate, area).ToBasicForm();
                        f.Terminate();
                        success = Lp_norm(r, new RealVector(sol)) < Tolerance;
                    }
                }
                resultsPerFunction[id] = success;
            }

            return resultsPerFunction.All(_ => _);
        }
        
    }
}